### PR TITLE
Feature/test ir341

### PR DESCRIPTION
**Pull request by luckychess**
## What is this pull request?
Changes related to crypto:
- hash provider removed from everywhere
- crypto reworked (hashes and signatures)
- everything uses crypto were changed accordingly
- changed model for proto - transaction and query: both of them now have a payload field contained all the data and signature field for a payload's signature
- all code is changed according to new model (serializers, deserializers, converters and some other)

Other changes:
- cmake proto compilation improvements
- queries in serializers are const now
- std::abs compilation issue (again!)
   
## Why do you implement it? Why do we need this pull request?
Now to use crypto functions you have to serialize your model object to proto object and use its string representation (see code example below). This way allows us to get consistent hashes/signatures which do not depend from target platform internal binary data representation (endianess and so on). Proto model changes simplify crypto usage: now you always have a payload field and hashes/signs are calculated from it.
  
## How to use the features provided in the pull request?
Typical workflow for hashing and signing now:
```
protocol::Transaction tx;
hash256_t hash = sha3_256(tx.payload().SerializeAsString());
iroha::sign(hash.to_string(), pubkey, privkey);
```

To get a signature from string there is from_string function available:
```
Signature sign{};
sign.pubkey = pubkey_t::from_string(pubkey_string);
sign.signature = sig_t::from_string(signature_string);
```

### CODE REVIEW

**Comment by muratovv** [comment]
No reason to start review before we are not discussing, how model will looks like. I guess better to provide draft of model interface and after it make changes in code. Short review demonstrate that code now very dirty: changes not obvious and come pieces looks horrible. Another problem is no reason to review 4+k loc changes, separate pr to atomic parts - new model, crypto, etc.

File: libs/builder/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,22 @@
+message (STATUS "Hi there")
```
Comment (neewy) [inline libs/builder/CMakeLists.txt]: privet

File: irohad/ametsuchi/block_query.hpp [hunk 2]
```diff
@@ -18,6 +18,8 @@
 #ifndef IROHA_BLOCK_QUERY_HPP
 #define IROHA_BLOCK_QUERY_HPP
 
+#include <cmath>
```
Comment (Warchant) [inline irohad/ametsuchi/block_query.hpp]: why is this here?

File: irohad/model/converters/impl/pb_query_factory.cpp [hunk 3]
```diff
@@ -30,91 +30,100 @@ namespace iroha {
       PbQueryFactory::PbQueryFactory() {
         log_ = logger::log("PbQueryFactory");
         serializers_[typeid(GetAccount)] = &PbQueryFactory::serializeGetAccount;
-        serializers_[typeid(GetAccountAssets)] = &PbQueryFactory::serializeGetAccountAssets;
-        serializers_[typeid(GetAccountTransactions)] = &PbQueryFactory::serializeGetAccountTransactions;
-        serializers_[typeid(GetAccountAssetTransactions)] = &PbQueryFactory::serializeGetAccountAssetTransactions;
-        serializers_[typeid(GetSignatories)] = &PbQueryFactory::serializeGetSignatories;
+        serializers_[typeid(GetAccountAssets)] =
+            &PbQueryFactory::serializeGetAccountAssets;
+        serializers_[typeid(GetAccountAssets)] =
```
Comment (Warchant) [inline irohad/model/converters/impl/pb_query_factory.cpp]: get account assets here is twice

File: libs/common/types.hpp [hunk 4]
```diff
@@ -80,6 +82,17 @@ namespace iroha {
       }
       return res;
     }
+
+    static blob_t<size_> from_string(const std::string &data) {
+      if(data.size() != size_) {
+        throw BadFormatException("blob_t: input string has incorrect length");
```
Comment (Warchant) [inline libs/common/types.hpp]: to other reviewers: this will be changed to optional as soon as Andrei will merge #552

File: irohad/model/CMakeLists.txt [hunk 5]
```diff
@@ -3,15 +3,26 @@ add_subdirectory(converters)
 
 add_library(model
     model_crypto_provider_impl.cpp
-    model_hash_provider_impl.cpp
     impl/model_operators.cpp
+    converters/impl/pb_block_factory.cpp
+    converters/impl/pb_transaction_factory.cpp
+    converters/impl/pb_command_factory.cpp
```
Comment (lebdron) [inline irohad/model/CMakeLists.txt]: Looks like something went wrong, since these files are in `*_model_converters` targets in `irohad/model/converters/CMakeLists.txt`.

File: libs/crypto/crypto.hpp [hunk 6]
```diff
@@ -63,8 +65,6 @@ namespace iroha {
    * @param seed
    * @return
    */
-  ed25519::keypair_t create_keypair(blob_t<32> seed);
-
-
+  keypair_t create_keypair(blob_t<32> seed);
```
Comment (l4l) [inline libs/crypto/crypto.hpp]: Maybe `create_keypair(void)` will be a better solution (no issues with seed)?

File: test/integration/client_test.cpp [hunk 7]
```diff
@@ -175,10 +175,13 @@ TEST_F(ClientServerTest, SendTxWhenStatelessInvalid) {
...
 
   auto doc = iroha::model::converters::stringToJson(json_tx).value();
-  iroha::model::converters::JsonTransactionFactory transactionFactory;
-  auto tx = transactionFactory.deserialize(doc).value();
-  iroha::model::HashProviderImpl hashProvider;
-  tx.tx_hash = hashProvider.get_hash(tx);
+  iroha::model::converters::JsonTransactionFactory jsonTransactionFactory;
+  auto tx = jsonTransactionFactory.deserialize(doc).value();
+  iroha::model::converters::PbTransactionFactory pbTransactionFactory;
+  auto hash = iroha::sha3_256(pbTransactionFactory.serialize(tx)
```
Comment (l4l) [inline test/integration/client_test.cpp]: That will definitely not a proper way to use crypto stuff. Add todo for writing subfunctions like `hash(const Transaction&)` or `hash(const Block&)`

File: irohad/ametsuchi/block_query.hpp (hunk 2 above)
Comment (luckychess) [reply irohad/ametsuchi/block_query.hpp]: Well, that's really rx-cpp issue and this include should be somewhere inside their code. This place is bad but other places are not really better. And yes, without this line some systems (at least, my one) get a compilation error about ambiguous call of std::abs somewhere inside rc-cpp internals.

File: libs/crypto/crypto.hpp (hunk 6 above)
Comment (luckychess) [reply libs/crypto/crypto.hpp]: We could create an overload for ```create_keypair(void)``` but do not remove a version with seed because we have 2 different ```create_seed()``` implementations now (and both are used). I can add it if you think it will be useful.

File: test/integration/client_test.cpp (hunk 7 above)
Comment (luckychess) [reply test/integration/client_test.cpp]: Good point. I added TODO; actual work will require a lot of code changes so it should be done in separate PR I think.

File: irohad/model/CMakeLists.txt (hunk 5 above)
Comment (luckychess) [reply irohad/model/CMakeLists.txt]: Thanks, fixed.

File: libs/crypto/crypto.hpp (hunk 6 above)
Comment (l4l) [reply libs/crypto/crypto.hpp]: Yep, don't need to be deleted and as long as pr related to reworking crypto imo this is fine to add

File: test/integration/client_test.cpp (hunk 7 above)
Comment (l4l) [reply test/integration/client_test.cpp]: Yes, thanks

File: test/module/irohad/peerservice/peerservice_test.cpp [hunk 8]
```diff
@@ -33,7 +33,7 @@ class Service : public testing::Test {
   Service() {
     std::vector<ed25519::keypair_t> kps(NPEERS);
     std::generate_n(kps.begin(), NPEERS,
-                    []() { return create_keypair(create_seed()); });
+                    []() { return create_keypair(); });
```
Comment (l4l) [inline test/module/irohad/peerservice/peerservice_test.cpp]: can be just `std::generate_n(kps.begin(), NPEERS, create_keypair)`
and btw `std::generate(kps.begin(), kps.end(), create_keypair)` seems less error-prone for me

File: libs/crypto/ed25519_impl.cpp [hunk 9]
```diff
@@ -81,4 +81,10 @@ namespace iroha {
 
     return keypair_t{.pubkey = pub, .privkey = priv};
   }
+
+  /**
+   * Create new keypair with a default seed (by create_seed())
```
Comment (l4l) [inline libs/crypto/ed25519_impl.cpp]: Imo comments in header is enough

File: irohad/ametsuchi/block_query.hpp (hunk 2 above)
Comment (l4l) [reply irohad/ametsuchi/block_query.hpp]: Same problem, anyway this include doesn't help

Comment (l4l) [reply irohad/ametsuchi/block_query.hpp]: Look at [this](https://gist.github.com/l4l/701b2d56dfc9b9714ab4955b1df16902) one

File: test/integration/client_test.cpp (hunk 7 above)
Comment (neewy) [reply test/integration/client_test.cpp]: @luckychess please tell me which TODO task to add to backlog 😄
